### ISSUE TITLE
[Drive] Autocached root_folder_id disables team drive access

### ISSUE THREAD

**Issue by wlerin** [rclone/rclone#4051]
#### What is the problem you are having with rclone?
Since ba7c2ac443233ec4b73eed1054bd68c59eb3d8dd, accessing a Google Drive remote adds a root_folder_id to the config file. Thereafter, thanks to 473a437163027ae6116b948a71d2de467890f2d2 , the `--drive-team-drive` flag (or equivalent environment variable) is ignored on that remote. I'm not entirely clear on the motivation behind the second change (nor have I used `--drive-shared-with-me`), but surely a manually specified team drive should take precedence over the autogenerated root folder id?

#### What is your rclone version (output from `rclone version`)
The issue occurs with 1.51.0 and 1.50.2. Rolling back to 1.50.1 fixes it.

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)
Ubuntu 18.04, 64 bit

####  Which cloud storage system are you using? (eg Google Drive)
Google Drive

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)
Any command using --drive-team-drive, after running any command without it once on the same remote. Any information returned, or actions performed, will be on the base remote rather than the team drive.

**Comment by ncw** [comment]
> Since [ba7c2ac](https://github.com/rclone/rclone/commit/ba7c2ac443233ec4b73eed1054bd68c59eb3d8dd), accessing a Google Drive remote adds a root_folder_id to the config file. Thereafter, thanks to [473a437](https://github.com/rclone/rclone/commit/473a437163027ae6116b948a71d2de467890f2d2) , the `--drive-team-drive` flag (or equivalent environment variable) is ignored on that remote.

Hmm, that was an unexpected consequence.

One of `team_drive` or `root_folder_id` has to win as they both specify the same thing 473a437163027ae6116b948a71d2de467890f2d2 made it so `root_folder_id` would win so you could then specify that in on the command line.

What you want is the reverse!

I note that you can work-around this with `--drive-root-folder-id ""`

I'm not sure the best thing to do... here are some ideas

- move the cached `root_folder_id` to a `cached_root_folder_id` and make it so only the user configures `root_folder_id`
    - this has a backwards compatibility problem as there are a lot of `root_folder_id` out there which rclone already set
- warn the user that both `team_drive` and `root_folder_id` are set
- error out if both `team_drive` and `root_folder_id` are set
- remove the caching of `root_folder_id` and accept the performance hit
    - again a backwards compat problem

Note that by the time the config has got to the backend, it can't tell whether stuff was set by command line or in the config file.

**Comment by wlerin** [comment]
Regardless of where it's set, in what situation would one want `root_folder_id` to take precedence over `team_drive`? If it is configured as a `team_drive`, they should be identical, no? If one is trying to use `--drive-shared-with-me` with a remote configured as a `team_drive` should it not just fail, or return a warning? What other commands would prefer the actual root rather than the team drive?

I would recommend option 2, but only when it matters (e.g. with `--drive-shared-with-me` which won't return anything when using a team drive), and otherwise ignore it, using the `team_drive` value over the `root_folder_id`. Otherwise a lot of unnecessary warnings will be generated. Option 3 would cause hard exits during normal operation (since `root_folder_id` is still set automatically).

**Comment by darthShadow** [comment]
I agree with @wlerin, I am not sure if it's worth the trouble to support both team_drive and root_folder_id simultaneously when they have different values, especially if the root_folder_id can be automatically set by rclone.

It should probably error out in such cases to avoid any issues with automatic syncs that users may be running.

**Comment by wlerin** [comment]
The main issue with erroring out is that doing so will cause hard exits in what I'd assume to be fairly common usage. Then again, this has been the case since Nov 19 and I'm the first one to report it that I can see--it wouldn't affect users who either aren't using team drives, or are making dedicated remotes for them.

Does the config file support global flags, like one to turn off automatically fetching and caching the root folder id for all remotes? Even just doing it per remote would be fine.

**Comment by bpoliakoff** [comment]
I'm considering using rclone  (using a service account with "domain wide delegation enabled) to migrate user data from a couple thousand local user home directories on a fileserver to many corresponding accounts in Google Drive.  As noted using `--drive-root-folder-id ""` is a workaround, but I just wanted developers to be aware of this use case scenario, rewriting the config file each time the command runs for a different user seems less than ideal.

**Comment by wlerin** [comment]
The other "workaround" is to always use `--drive-root-folder-id` instead of `--drive-team-drive`. It's a bit more verbose but as far as I can tell there's no difference in using one or the other, other than the former can point to individual folders as well.

**Comment by wlerin** [comment]
The "workaround" I posted above differs in at least one important case. The following two commands return different information for the same team drive ID (here given as XXXXX):

```
rclone size --drive-root-folder-id XXXXX drive:
rclone size --drive-team-drive XXXXX drive:
```

The root-folder-id version returns quickly and the result makes no sense (much too small).

**Comment by ncw** [comment]
@wlerin the `--drive-team-drive` flag also turns on the team drive flag within rclone which is needed to read team drives correctly.

Which version of rclone are you using? Have you tried [the latest beta](https://beta.rclone.org) as I've made several changes in this area.

**Comment by wlerin** [comment]
@ncw I have not tried the beta, in fact when I became aware of this issue I rolled all my machines back to v1.50.1. I'll have to take a look at the changes.

**Comment by ncw** [comment]
OK I've had a go at fixing this here

[v1.52.2-201-gdcbbe4b8-fix-4051-teamdrive-beta](https://beta.rclone.org/branch/v1.52.2-201-gdcbbe4b8-fix-4051-teamdrive-beta/) on branch [fix-4051-teamdrive](https://github.com/rclone/rclone/tree/fix-4051-teamdrive) (uploaded in 15-30 mins)

This uses the same technique I used in 5c5ad622087b5c0ec4f0a72f8fd21de6f18c9cb8 - namely ignoring the root_folder_id if team_drive is set but outputing a NOTICE level log if both are set.

I think it is unlikely people have root_folder_id and team_drive set deliberately so I think this will be a pragmatic compromise.

Testing appreciated :-)

If this works OK I'll stick it in 1.52.3

**Comment by ncw** [comment]
This is fixed now - all caching disabled in f50ab981f788d7fb0c507758857d4a190d3d8fb7

This has been released in v1.52.3
